lib/package.py

```diff
@@ -101,8 +101,31 @@ def __init__(self, name, force_rebuild=True):
             raise exception.PackageDescriptorError(
                 "Failed to find %s's YAML descriptor" % self.name)
 
+        # load package metadata YAML file
         self._load()
 
+        # get global config information which may override YAML
+        packages = CONF.get('build_packages').get('packages') or []
+        for package in packages:
+            if package.startswith(name):
+                if package == name:
+                    break
+                else:
+                    package_parts = package.split("#")
+                    # assume that first source is the main one and it is always
+                    # desired to override it
+                    source_type = self.sources[0].keys()[0]
+                    if package_parts[1]:
+                        self.sources[0][source_type]["src"] = package_parts[1]
+                    if len(package_parts) == 4:
+                        if package_parts[2]:
+                           self.sources[0][source_type]["branch"] = package_parts[2]
+                        if package_parts[3]:
+                           self.sources[0][source_type]["commit_id"] = package_parts[3]
+                    elif len(package_parts) == 3:
+                        if package_parts[2]:
+                           self.sources[0][source_type]["commit_id"] = package_parts[2]
```

Comment:
I find it a little hard to follow the code with the nested ifs and indexes everywhere.
I know you're not a fan of EAFP, but here's a suggestion. The try block stops when it finds an undefined variable. Feel free to keep as is if you dislike it.
```
package_parts = package.split("#")
if len(package_parts) == 1:
    break
elif len(package_parts) == 2:
    _, src = package_parts
elif len(package_parts) == 3:
    _, src, commit_id = package_parts
elif len(package_parts) == 4:
    _, src, branch, commit_id = package_parts
else:
    raise "Too many parts"
try:
    if src:
       self.sources[0][source_type]["src"] = src
    if commit_id:
       self.sources[0][source_type]["commit_id"] = commit_id
    if branch:
       self.sources[0][source_type]["branch"] = branch
except:
    pass
```